controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -70,6 +95,82 @@ ZKStream newStream(final String scope, final String name) {
+        synchronized (lock) {
+            // Ensure that only one background refresh is happening. For this we will reference the future in refreshFutureRef
+            // If reference future ref is not null, we will return the reference to that future.
+            // It is set to null when refresh completes.
+            refreshFuture = this.refreshFutureRef;
+            if (this.refreshFutureRef == null) {
+                // no ongoing refresh, check if refresh is still needed
+                if (counter.get().compareTo(limit.get()) >= 0) {
+                    log.info("refreshing counter range. Current counter is {}. Current limit is {}", counter.get(), limit.get());
+
+                    // Need to refresh counter and limit. Start a new refresh future. We are under lock so no other
+                    // concurrent thread can start the refresh future.
+                    refreshFutureRef = getRefreshFuture()
+                            .exceptionally(e -> {
+                                // if any exception is thrown here, we would want to reset refresh future so that it can be retried.
+                                synchronized (lock) {
+                                    refreshFutureRef = null;
+                                }
+                                log.warn("Exception thrown while trying to refresh transaction counter range", e);
+                                throw new CompletionException(e);
+                            });
+                    // Note: refreshFutureRef is reset to null under the lock, and since we have the lock in this thread
+                    // until we release it, refresh future ref cannot be reset to null. So we will always return a non-null
+                    // future from here.
+                    refreshFuture = refreshFutureRef;
+                } else {
+                    // nothing to do
+                    refreshFuture = CompletableFuture.completedFuture(null);
+                }
+            }
+        }
+        return refreshFuture;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> getRefreshFuture() {
+        return storeHelper.createZNodeIfNotExist(ZKStoreHelper.COUNTER_PATH, BigLong.ZERO.toBytes())
+                .thenCompose(v -> storeHelper.getData(ZKStoreHelper.COUNTER_PATH)
+                        .thenCompose(data -> {
+                            BigLong previous = BigLong.fromBytes(data.getData());
+                            BigLong nextLimit = previous.add(previous, COUNTER_RANGE);
+                            return storeHelper.setData(ZKStoreHelper.COUNTER_PATH, new Data<>(nextLimit.toBytes(), data.getVersion()))
+                                    .thenAccept(x -> {
+                                        // Received new range, we should reset the counter and limit under the lock
+                                        // and then reset refreshfutureref to null
+                                        synchronized (lock) {
```

Comment:
`refreshRangeIfNeeded()` acquires `lock` and then invokes `getRefreshFuture()` which again acquires `lock`. Is it guaranteed that `thenAccept()`/ `exceptionally()` use the same thread? if yes, it implies that the thread is blocked  i.e. it is not a async operation. 

```
Tried the following example...
    public static void main(String[] args) throws Exception {
        Executor executor = Executors.newSingleThreadScheduledExecutor();

        System.out.println(Thread.currentThread().getId() + ": Main thread");

        CompletableFuture<Void> f = new CompletableFuture<>();

        executor.execute(() -> {
            try {
                System.out.println(Thread.currentThread().getId() + " START " );
                f.thenAccept( v -> {
                    System.out.println(Thread.currentThread().getId() + ": Accept!");
                });
                Thread.sleep(1000);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });

        f.complete(null); //Option 1.
        Thread.sleep(3000);
        f.thenCompose(v -> {
            System.out.println(Thread.currentThread().getId() + ": thenCompose!");
            return new CompletableFuture<>();
        });
//        f.complete(null); //Option 2
        Thread.sleep(3000);
    }

Option-1 Output:

1: Main thread
9 START 
9: Accept!
1: thenCompose!

Option-2 output:
1: Main thread
9 START 
1: thenCompose!
1: Accept!
```

Reply:
> Is it guaranteed that thenAccept()/ exceptionally() use the same thread? if yes
No there is no guarantee of the sort. 
The business logic assumes they can run concurrently on different threads. Hence locking at both the places. 
That is why we dont do any blocking operations while holding the lock. 
at an abstract level here are three places where locks are taken and the work we do under the lock (note: only local variable assignments, no blocking):
1. ```method refreshRangeIfNeeded() {
lock
futureRef = submit some work and get its future
release lock
}
```
2. ```method getRefreshFuture() {
future.thenaccept(result -> {
lock
update variables counter limit and futureRef
release lock
}
}```

3. ```getRefreshFuture.exceptionally(e -> {
lock
refreshFuture = null
release lock
}```